src/category_theory/sites/sheaf.lean

```diff
@@ -0,0 +1,830 @@
+lemma is_separated_for_and_exists_is_amalgamation_iff_sheaf_for :
+  is_separated_for P R ∧ (∀ (x : family_of_elements P R), x.compatible → ∃ t, x.is_amalgamation t) ↔
+  is_sheaf_for P R :=
+begin
+  rw [is_separated_for, ←forall_and_distrib],
+  apply forall_congr,
+  intro x,
+  split,
+  { intros z hx, exact exists_unique_of_exists_of_unique (z.2 hx) z.1 },
+  { intros h,
+    refine ⟨_, (exists_of_exists_unique ∘ h)⟩,
+    intros t₁ t₂ ht₁ ht₂,
+    apply (h _).unique ht₁ ht₂,
+    exact is_compatible_of_exists_amalgamation x ⟨_, ht₂⟩ }
+end
+
+/-- If `P` is separated for `R` and every family has an amalgamation is a sheaf. -/
+lemma is_separated_for.is_sheaf_for (t : is_separated_for P R) :
+  (∀ (x : family_of_elements P R), x.compatible → ∃ t, x.is_amalgamation t) →
+  is_sheaf_for P R :=
+begin
+  rw ← is_separated_for_and_exists_is_amalgamation_iff_sheaf_for,
+  apply and.intro t,
+end
+
+/-- Get the amalgamation of the given compatible family, provided we have a sheaf. -/
+noncomputable def is_sheaf_for.amalgamate
+  (t : is_sheaf_for P R) (x : family_of_elements P R) (hx : x.compatible) :
+  P.obj (op X) :=
+classical.some (t x hx).exists
+
+lemma is_sheaf_for.is_amalgamation
+  (t : is_sheaf_for P R) {x : family_of_elements P R} (hx : x.compatible) :
+  x.is_amalgamation (t.amalgamate x hx) :=
+classical.some_spec (t x hx).exists
+
+@[simp]
+lemma is_sheaf_for.valid_glue
+  (t : is_sheaf_for P R) {x : family_of_elements P R} (hx : x.compatible) (f : Y ⟶ X) (Hf : R f) :
+  P.map f.op (t.amalgamate x hx) = x f Hf :=
+t.is_amalgamation hx f Hf
+
+/-- If `P` is a sheaf for `R`, it is separated for `R`. -/
+lemma is_sheaf_for.is_separated_for : is_sheaf_for P R → is_separated_for P R :=
+λ q, (is_separated_for_and_exists_is_amalgamation_iff_sheaf_for.2 q).1
+
+/-- C2.1.3 in [Elephant] -/
+lemma is_sheaf_for_iff_generate :
+  is_sheaf_for P R ↔ is_sheaf_for P (generate R) :=
+begin
+  rw ← is_separated_for_and_exists_is_amalgamation_iff_sheaf_for,
+  rw ← is_separated_for_and_exists_is_amalgamation_iff_sheaf_for,
+  rw ← is_separated_for_iff_generate,
+  apply and_congr (iff.refl _),
+  split,
+  { intros q x hx,
+    apply exists_imp_exists _ (q _ (hx.restrict (le_generate R))),
+    intros t ht,
+    simpa [hx] using is_amalgamation_extend _ _ ht },
+  { intros q x hx,
+    apply exists_imp_exists _ (q _ (hx.sieve_extend _)),
+    intros t ht,
+    simpa [hx] using is_amalgamation_restrict (le_generate R) _ _ ht },
+end
+
+/--
+Every presheaf is a sheaf for the family {𝟙 X}.
+
+[Elephant] C2.1.5(i)
+-/
+lemma is_sheaf_for_singleton_iso (P : Cᵒᵖ ⥤ Type v) :
+  is_sheaf_for P (presieve.singleton (𝟙 X)) :=
+begin
+  intros x hx,
+  refine ⟨x _ (presieve.singleton_self _), _, _⟩,
+  { rintro _ _ ⟨rfl, rfl⟩,
+    simp },
+  { intros t ht,
+    simpa using ht _ (presieve.singleton_self _) }
+end
+
+/--
+Every presheaf is a sheaf for the maximal sieve.
+
+[Elephant] C2.1.5(ii)
+-/
+lemma is_sheaf_for_top_sieve (P : Cᵒᵖ ⥤ Type v) :
```

Comment:
```suggestion
lemma is_sheaf_for_top_presieve (P : Cᵒᵖ ⥤ Type v) :
```

Reply:
I think the name `sieve` is better here: the point is that we are taking the top sieve, coercing it to a presieve and testing the sheaf condition there. In particular, it is not necessarily the top presieve.